### PR TITLE
feat: option to add custom meta tags to the UI container

### PR DESCRIPTION
**Pull request by pritamsoni-hsr**
## What does this PR do?

feature to allow users to add custom meta tags directly to the lightning app.

under the hood, it writes these tags to `index.html` as the last step in the LightningApp constructor.

# Limitation

if a user runs multiple apps on a single machine at the same time the latest tags will be visible in all the apps.

image should be a valid url, rn users cannot provide a local filesystem path.

need to update docs 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/lightning_app/core/app.py [hunk 1]
```diff
@@ -133,6 +136,9 @@ def __init__(
 
         logger.debug(f"ENV: {os.environ}")
 
+        # update index.html
```
Comment (alecmerdler) [inline src/lightning_app/core/app.py:139]: Make a comment on why this should be in the constructor (it should only be written once during the app lifecycle).

**Review by alecmerdler — APPROVED** [review]

File: requirements/app/base.txt [hunk 2]
```diff
@@ -8,3 +8,4 @@ croniter>=1.3.0, <1.4.0  # strict; TODO: for now until we find something more ro
 traitlets>=5.3.0, <=5.4.0
 arrow>=1.2.0, <1.2.4
 lightning-utilities==0.3.*
+beautifulsoup4==4.8.2
```
Comment (tchaton) [inline requirements/app/base.txt:11]: Why this is needed? Let's avoid adding new dependencies.

File: src/lightning_app/utilities/add_meta.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+    with original_file.open() as f:
+        original = f.read()
+
+    with entry_file.open("w") as f:
+        f.write(_get_updated_content(original=original, info=info))
+
+
+def _get_updated_content(original: str, info: AppInfo) -> str:
+    soup = BeautifulSoup(original, "html.parser")
```
[the remark this answers is not in the document]
Comment (pritamsoni-hsr) [reply src/lightning_app/utilities/add_meta.py:43]: with a parser it's very much clean, we can use regex but there is a high chance that we'll miss some edge cases, that's why a parser is good to use here.

also, we'll soon be expanding extra things that users can pass.

[the remark this answers is not in the document]
Comment (tchaton) [reply src/lightning_app/utilities/add_meta.py:43]: Not a huge fan as any extra dependency makes maintainability harder.

Fine for now.

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

**Review by kaushikb11 — APPROVED** [review]
LGTM!

File: requirements/app/base.txt (hunk 2 above)
Comment (awaelchli) [inline requirements/app/base.txt:11]:
```suggestion
beautifulsoup4<=4.8.2
```
It's important we don't pin. The pip dependency resolver is not exactly flexible. It is unlikely that for the changes here you require the exact version on the dot, and no other one works.

File: src/lightning_app/utilities/add_meta.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
+from dataclasses import dataclass
+from typing import List, Optional
+
+from bs4 import BeautifulSoup
+
+
+@dataclass
+class AppInfo:
```
Comment (awaelchli) [inline src/lightning_app/utilities/add_meta.py:8]: Originally, the idea was to have that in the .lightning file that is stored alongside the app. Today we save only the name of the app there. 

Are we sure about splitting this up?

File: src/lightning_app/core/app.py [hunk 5]
```diff
@@ -46,6 +48,7 @@ def __init__(
         self,
         root: "lightning_app.LightningFlow",
         debug: bool = False,
+        info: add_meta.AppInfo = None,
```
Comment (awaelchli) [inline src/lightning_app/core/app.py:51]: This argument is missing in the docstring.

**Review by awaelchli — COMMENTED** [review]
I share same concerns with Thomas about the dependency. Have we looked at the code, are we happy with how it is maintained, how heavy/light it is etc?

File: src/lightning_app/core/app.py (hunk 5 above)
Comment (nohalon) [reply src/lightning_app/core/app.py:51]: yeah @pritamsoni-hsr can you add `into` to the arguments on line 68

File: src/lightning_app/utilities/add_meta.py (hunk 4 above)
Comment (manskx) [reply src/lightning_app/utilities/add_meta.py:8]: Agree with @awaelchli, we already have `AppConfig` object: https://github.com/Lightning-AI/lightning/blob/7dbd038a9455f3c67da3fe19ef579058feef51b2/src/lightning_app/utilities/packaging/app_config.py#L13 which is the `.lightning` file and can be used to contain configurations or meta data of the app

File: src/lightning_app/core/app.py [hunk 6]
```diff
@@ -18,12 +18,14 @@
...
     FLOW_DURATION_SAMPLES,
     FLOW_DURATION_THRESHOLD,
+    FRONTEND_DIR,
     STATE_ACCUMULATE_WAIT,
 )
 from lightning_app.core.queues import BaseQueue, SingleProcessQueue
 from lightning_app.frontend import Frontend
 from lightning_app.storage import Drive, Path
 from lightning_app.storage.path import storage_root_dir
+from lightning_app.utilities import add_meta
```
Comment (manskx) [inline src/lightning_app/core/app.py:28]: Maybe `add_meta` can be renamed to something else?

File: src/lightning_app/utilities/add_meta.py (hunk 4 above)
Comment (pritamsoni-hsr) [reply src/lightning_app/utilities/add_meta.py:8]: this sort of API requirement came from exec team.

File: src/lightning_app/core/app.py [hunk 7]
```diff
@@ -133,6 +136,10 @@ def __init__(
 
         logger.debug(f"ENV: {os.environ}")
 
+        # update index.html,
+        # this should happen once for all apps before the ui server starts running.
+        add_meta.update_index_file_with_info(FRONTEND_DIR, info=info)
```
Comment (manskx) [inline src/lightning_app/core/app.py:141]: This means that the default `index.html` will be changed for all apps, is this intended?

File: src/lightning_app/core/app.py (hunk 5 above)
Comment (pritamsoni-hsr) [reply src/lightning_app/core/app.py]: yeah I added this now and marked this in PR description.

File: src/lightning_app/core/app.py (hunk 6 above)
Comment (pritamsoni-hsr) [reply src/lightning_app/core/app.py]: frontend.py

**Comment by pritamsoni-hsr** [comment]
bs4 is around 2MB in size, @awaelchli. I think we should just use it instead of writing parsers from scratch.

File: src/lightning_app/core/app.py (hunk 7 above)
Comment (manskx) [reply src/lightning_app/core/app.py]: Imagine these two apps:
app_1.py
```python

app = L.LightningApp(info=AppInfo(name="Im a cool app"))

```
the title of the HTML will be "Im a cool app"

app_2.py
```python

app = L.LightningApp()

```

if you run app_1 first then app_2, will app_2 have "Im a cool app" title ?

**Review by tchaton — APPROVED** [review]
Approving to unblock releasing!

File: src/lightning_app/core/app.py (hunk 7 above)
Comment (pritamsoni-hsr) [reply src/lightning_app/core/app.py]: no, it'll be reverted to original.

**Pull request merged**